### DISCUSSION TITLE
A way to load a script for single page?

### DISCUSSION THREAD

**Discussion by gasatrya** [11ty/eleventy#2506] [Q&A] [answered]
Is there a way to load a js file for just a single page? Let's say I have a `generator.js` and I just want to load this file only on **Generator** page.

Thanks!

**Comment by cfjedimaster** [comment, marked as answer]
The simplest solution is to just add the script tag. If you are using Markdown templates, nothing stops you from using HTML, nor any of the other ones. Eleventy lets you output any HTML you want.

Now if you want to get fancier and ensure it gets loaded in the HEAD part of your doc and you are using an Eleventy layout template, use front matter to specify something, have the layout template look for it and optionally load it. For example:

```
---
title: Generator page
script: generator.js
---
```

And in your layout, if script is defined, you can output `<script src="{{ script }}"></script>`.

**Reply by pdehaan** [reply]
Another option is to use [YAML multiline strings](https://yaml-multiline.info/) and define your optional script tag(s) in the template (note the confusingly peculiar `>` in the YAML front matter):

```njk
---
# src/index.njk
title: Issue 2205
layout: layouts/base.njk
scripts: >
  <script src="/generator.js" defer></script>
---

Hello the world!!
````

And then just output the [optional] `{{ scripts }}` variable in your layout file wherever you want the script tags to appear (like just before the `</body>` tag):

```njk
---
# src/_includes/layouts/base.njk
---
<!DOCTYPE html>
<html lang="en">
<head>
  <title>{{ title }}</title>
</head>
<body>

  <main>{{ content | safe }}</main>
  {{ scripts | safe }}

</body>
</html>
```

Or, if you want to _really_ go overboard you could use [Nunjucks `{% block %}` tags](https://mozilla.github.io/nunjucks/templating.html#block), or try [LiquidJS `{% block %}` and `{% layout %}` tags](https://liquidjs.com/tags/layout.html). Although, I don't think I've tried the LiquidJS version w/ Eleventy before.
But this would let you do stuff like `{% block header %}...{% endblock %}` and `{% block content %}...{% endblock %}`, etc. if you wanted to customize a lot of content per page (and where the scripts appear, if you want them in the `<head>` or `<body>`.

**Reply by gasatrya** [reply]
Thank you for the answer for both of you. I'll try which one is works for me.

**Comment by btrem** [comment]
I do something similar to what @pdehaan describes in https://github.com/11ty/eleventy/discussions/2506#discussioncomment-3223619. But my solution allows for easier to write and read front matter.

In `base.njk` I have the following code in my `<head>`:

    {% block script %}
        {% if script %}
            <script src="/javascript/{{ script }}"></script>
        {% endif %}
    {% endblock %}

Then, in any `.md` file, I can add a script via front matter:

    ---
    layout: base.njk
    permalink: foo.html
    script: myscript.js
    ---

And `foo.html` will have the following code:

        <script src="/javascript/myscript.js"></script>

**Comment by tomrsvr** [comment]
My approach is to use an if statement in my base file linking to a Nunjucks partial that contains the full script that needs to be show within the head. Unsure if it works well for larger use cases, but for getting a specific script into the head of a specific page (where title comes from frontmatter), it works well:

```
 {% if title == 'contact' %}
    {% include "script-name.njk" %}
 {% endif %}
```

**Comment by soska** [comment]
Hi! What would some one do if they wanted to preprocess the script in question? Like writing it in typescript, for example?
